### PR TITLE
KAFKA-14097: make producer ID expiration a dynamic config

### PR DESCRIPTION
**Pull request by jolshan**
Changed the integer value for producer ID expiration to a dynamically configurable value.

Tested that the configuration can be changed dynamically and works as expected.

I considered adding the producer expiration check interval into the config, but since it is not dynamic (and is harder to make it dynamic), I decided to keep it separate. A followup could be to incorporate all the producer state manager configs into this new object.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/log/ProducerStateManager.scala [hunk 1]
```diff
@@ -939,3 +940,24 @@ object SnapshotFile {
...
+  }
+
+  override def validateReconfiguration(newConfig: KafkaConfig): Unit = {
+    if (newConfig.producerIdExpirationMs < 0)
+      throw new ConfigException(s"${KafkaConfig.ProducerIdExpirationMsProp} cannot be less than 0, current value is $producerIdExpirationMs")
+  }
+}
+
+object ProducerStateManagerConfig {
+  val ReconfigurableConfigs = Set(KafkaConfig.ProducerIdExpirationMsProp)
```
Comment (dajac) [inline core/src/main/scala/kafka/log/ProducerStateManager.scala:962]: Is it used anywhere? It does not seem to be.

**Comment by dajac** [comment]
@jolshan Left a small comment. Could you rebase it as well?

File: core/src/main/scala/kafka/log/ProducerStateManager.scala (hunk 1 above)
Comment (jolshan) [reply core/src/main/scala/kafka/log/ProducerStateManager.scala:962]: This is required to override the class.

Comment (jolshan) [reply core/src/main/scala/kafka/log/ProducerStateManager.scala:962]: It is also used in core/src/main/scala/kafka/server/DynamicBrokerConfig.scala

File: core/src/main/scala/kafka/log/ProducerStateManager.scala [hunk 2]
```diff
@@ -939,3 +940,24 @@ object SnapshotFile {
     SnapshotFile(file, offset)
   }
 }
+
+class ProducerStateManagerConfig(@volatile var producerIdExpirationMs: Int) extends Logging with BrokerReconfigurable {
+
+  override def reconfigurableConfigs: Set[String] = Set(KafkaConfig.ProducerIdExpirationMsProp)
```
Comment (dajac) [inline core/src/main/scala/kafka/log/ProducerStateManager.scala]: Should we use `ProducerStateManagerConfig.ReconfigurableConfigs` to be consistent?

**Review by dajac — APPROVED** [review]
LGTM, thanks.

**Pull request merged**